src/app/reporting/LoggingManagement.cpp

```diff
@@ -0,0 +1,979 @@
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "LoggingManagement.h"
+#include "LoggingConfiguration.h"
+#include <app/InteractionModelEngine.h>
+#include <app/MessageDef/EventDataElement.h>
+#include <core/CHIPEventLoggingConfig.h>
+#include <core/CHIPTLVUtilities.hpp>
+#include <inttypes.h>
+#include <support/CodeUtils.h>
+#include <support/ErrorStr.h>
+#include <support/logging/CHIPLogging.h>
+#include <system/SystemTimer.h>
+
+using namespace chip::TLV;
+
+namespace chip {
+namespace Platform {
+// TODO: Implement CriticalSectionEnter per platform
+void CriticalSectionEnter() {}
+
+void CriticalSectionExit() {}
+} // namespace Platform
+} // namespace chip
+
+namespace chip {
+namespace app {
+namespace reporting {
+
+static LoggingManagement sInstance;
+
+/**
+ * @brief
+ *   A TLVReader backed by CircularEventBuffer
+ */
+class CircularEventReader : public chip::TLV::TLVReader
+{
+public:
+    /**
+     * @brief
+     *   Initializes a TLVReader object backed by CircularEventBuffer
+     *
+     * Reading begins in the CircularTLVBuffer belonging to this
+     * CircularEventBuffer.  When the reader runs out of data, it begins
+     * to read from the previous CircularEventBuffer.
+     *
+     * @param[in] apBuf A pointer to a fully initialized CircularEventBuffer
+     *
+     */
+    void Init(chip::app::reporting::CircularEventBufferWrapper * apBuf);
+
+    virtual ~CircularEventReader() = default;
+};
+
+LoggingManagement & LoggingManagement::GetInstance(void)
+{
+    return sInstance;
+}
+
+struct ReclaimEventCtx
+{
+    CircularEventBuffer * mpEventBuffer;
+    size_t mSpaceNeededForEvent;
+};
+
+/**
+ * @brief
+ *  Internal structure for traversing event list.
+ */
+struct CopyAndAdjustDeltaTimeContext
+{
+    CopyAndAdjustDeltaTimeContext(chip::TLV::TLVWriter * aWriter, EventLoadOutContext * inContext);
+
+    chip::TLV::TLVWriter * mpWriter;
+    EventLoadOutContext * mpContext;
+};
+
+/**
+ * @brief
+ *  Internal structure for traversing events.
+ */
+struct EventEnvelopeContext
+{
+    EventEnvelopeContext(void);
+
+    size_t mNumFieldsToRead;
+    uint32_t mDeltaTime;
+    int64_t mDeltaUtc;
+    PriorityLevel mPriority;
+};
+
+CHIP_ERROR LoggingManagement::AlwaysFail(chip::TLV::CHIPCircularTLVBuffer & apBuffer, void * apAppData,
+                                         chip::TLV::TLVReader & inReader)
+{
+    return CHIP_ERROR_NO_MEMORY;
+}
+
+CHIP_ERROR LoggingManagement::CopyToNextBuffer(CircularEventBuffer * apEventBuffer)
+{
+    CircularTLVWriter writer;
+    CircularTLVReader reader;
+    CircularEventBuffer checkpoint   = *(apEventBuffer->mpNext);
+    CircularEventBuffer * nextBuffer = apEventBuffer->mpNext;
+    CHIP_ERROR err;
+
+    // Set up the next buffer s.t. it fails if needs to evict an element
+    nextBuffer->mProcessEvictedElement = AlwaysFail;
+
+    writer.Init(*nextBuffer);
```

Comment:
can these initializers return CHIP_ERROR? I am unused to Init without error checking.